app/modules/websocket.c

```diff
@@ -0,0 +1,332 @@
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    return luaL_error(L, "Client websocket is nil.\n");
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (lua_type(L, 2) != LUA_TNIL && lua_type(L, 2) != LUA_TFUNCTION && lua_type(L, 2) != LUA_TLIGHTFUNCTION) {
+    return luaL_error(L, "bad argument #2 to 'onClose' (function or nil expected, got something else)");
+  }
+
+  if (data->onClose != LUA_NOREF) {
+    luaL_unref(L, LUA_REGISTRYINDEX, data->onClose);
+  }
+
+  if (lua_type(L, 2) == LUA_TNIL) {
+    data->onClose = LUA_NOREF;
+  } else {
+    lua_pushvalue(L, 2);  // copy argument (func) to the top of stack
+    data->onClose = luaL_ref(L, LUA_REGISTRYINDEX);
+  }
+
+  return 0;
+}
+
+static int websocketclient_connect(lua_State *L) {
+  NODE_DBG("websocketclient_connect is called.\n");
+
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+  luaL_argcheck(L, ws, 1, "Client websocket expected");
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    return luaL_error(L, "Client websocket is nil.\n");
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (ws->connectionState != 0 && ws->connectionState != 4) {
+    return luaL_error(L, "Websocket already connecting or connected.\n");
+  }
+  ws->connectionState = 0;
+
+  lua_pushvalue(L, 1);  // copy userdata to the top of stack to allow ref
+  data->self_ref = luaL_ref(L, LUA_REGISTRYINDEX);
+
+  const char *url = luaL_checkstring(L, 2);
+  ws_connect(ws, url);
+
+  return 0;
+}
+
+static int websocketclient_send(lua_State *L) {
+  NODE_DBG("websocketclient_send is called.\n");
+
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+  luaL_argcheck(L, ws, 1, "Client websocket expected");
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    return luaL_error(L, "Client websocket is nil.\n");
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (ws->connectionState != 3) {
+    // should this be an onFailure callback instead?
+    return luaL_error(L, "Websocket isn't connected.\n");
+  }
+
+  int msgLength;
+  const char *msg = luaL_checklstring(L, 2, &msgLength);
+
+  int opCode = 1; // default: text message
+  if (lua_gettop(L) == 3) {
+    opCode = luaL_checkint(L, 3);
+  }
+
+  ws_send(ws, opCode, msg, (unsigned short) msgLength);
+  return 0;
+}
+
+static int websocketclient_close(lua_State *L) {
+  NODE_DBG("websocketclient_close.\n");
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+
+  ws_close(ws);
+  return 0;
+}
+
+static int websocketclient_gc(lua_State *L) {
+  NODE_DBG("websocketclient_gc\n");
+
+  ws_info *ws = (ws_info *) luaL_checkudata(L, 1, METATABLE_WSCLIENT);  
+  luaL_argcheck(L, ws, 1, "Client websocket expected");
+
+  if (ws == NULL || ws->reservedData == NULL) {
+    return luaL_error(L, "Client websocket is nil.\n");
+  }
+  ws_data *data = (ws_data *) ws->reservedData;
+
+  if (data->onConnection != LUA_NOREF) {
```

Comment:
Check for `LUA_NOREF` is redundant since [`luaL_unref()`](http://pgl.yoyo.org/luai/i/luaL_unref) handles this case.